src/FieldSet/styles.js

```diff
@@ -0,0 +1,49 @@
+import css from 'styled-jsx/css'
+
+import { spacers, theme } from '../theme.js'
+
+export default css`
+    fieldset {
+        border: none;
+        margin: 0;
+        padding: 0;
+    }
+    legend {
+        font-size: 16px;
+        color: ${theme.default};
+        margin: 0 0 ${spacers.dp12};
+    }
+
+    legend.required::after {
+        padding-left: ${spacers.dp4};
+        content: '*';
+    }
+
+    legend.valid {
+        color: ${theme.valid};
+    }
+
+    legend.warning {
+        color: ${theme.warning};
+    }
+
+    legend.error {
+        color: ${theme.error};
+    }
+    div {
+        display: flex;
+        flex-direction: row;
+    }
+
+    div.stacked {
+        flex-direction: column;
+    }
+
+    div > :global(.fieldset-input) {
```

Comment:
It is awkward that the `FieldSet` adds spacing to form components. 

If we have some components within a `FieldSet` and some outside, the fields are going to have different margins solely based on the position in the hierarchy.

Reply:
It could also happen the, due to a different specificity, the margin applied here will be ignored.
(I didn't think about that yesterday, when @HendrikThePendric and I were talking about this)

But specificity is another good reason not to interfere with a component's style but wrap it with another element instead (of course we could still have collapsing margins, so if the inner element has a bigger margin, that one will win, be we won't have 0-margin values anymore)

Reply:
Although the name [`FormControl`](https://github.com/dhis2/ui-core/blob/master/src/FormControl/index.js) might be a bit clunky, the function of it is perfect for this. `Field` would be an excellent name:

```
  <Field>
    <Input
      label="text"
    />
    <Help>text</Help>
  </Field>
```

Reply:
I think that component would be an internal wrapper inside `FieldSet` anyways,
so I have no objections to `Field`, but I think `FieldWrap` is more accurate.

Reply:
> 
> 
> I think that component would be an internal wrapper inside `FieldSet` anyways,
> so I have no objections to `Field`, but I think `FieldWrap` is more accurate.

That doesn't address the problem that components inside `FieldSet` are rendered different solely due to the fact that they are put inside a `FieldSet` component:

![image](https://user-images.githubusercontent.com/185449/63748528-d16b2900-c8a9-11e9-80d4-5a7949707dc1.png)

```js-jsx
<>
    <Input value="This value is valid" valid label="A valid thing" />
    <Help valid>Foobar</Help>

    <FieldSet label="Choose something">
        <Radio
            onChange={onChange}
            name="radio"
            value="first"
            label="First"
        />
        <Radio
            onChange={onChange}
            name="radio"
            value="second"
            label="Second"
            checked
        />
        <Radio
            onChange={onChange}
            name="radio"
            value="third"
            label="Third"
        />
    </FieldSet>
    <Input value="This value is invalid" error label="An invalid thing" />
    <Help valid>Foobar</Help>
</>
```

We could wrap every input in a `FieldSet`:

```js-jsx
<FieldSet>
    <Input value="This value is invalid" error label="An invalid thing" />
    <Help valid>Foobar</Help>
</FieldSet>
```

But that looks a bit funky unless we add `stacked`:

![image](https://user-images.githubusercontent.com/185449/63748721-2c9d1b80-c8aa-11e9-8551-5e79f0df7b4e.png)

With `stacked`:

![image](https://user-images.githubusercontent.com/185449/63748765-45a5cc80-c8aa-11e9-85bd-afadbd577b11.png)

Still not perfect...

If we use the `FormControl` component though:

![image](https://user-images.githubusercontent.com/185449/63749303-6fabbe80-c8ab-11e9-89f5-3d5345e3ec62.png)

```js-jsx
<>
    <FormControl>
        <Input value="This value is valid" valid label="A valid thing" />
        <Help valid>Foobar</Help>
    </FormControl>

    <FieldSet label="Choose something" stacked>
        <FormControl>
            <Radio
                onChange={onChange}
                name="radio"
                value="first"
                label="First"
            />
        </FormControl>

        <FormControl>
            <Radio
                onChange={onChange}
                name="radio"
                value="second"
                label="Second"
                checked
            />
        </FormControl>

        <FormControl>
            <Radio
                onChange={onChange}
                name="radio"
                value="third"
                label="Third"
            />
        </FormControl>
    </FieldSet>

    <FormControl>
        <Input value="This value is invalid" error label="An invalid thing" />
        <Help valid>Foobar</Help>
    </FormControl>
</>
```

If we go from `stacked` on the group of radio controls to `inline`, then the `margin-right` is gone too, so that would need to be sorted.

It is very explicit, it gets rid of the `React.Children.map` and `React.cloneElement` and just renders the `children` as they are defined.

These are very low-level components, and will not be used directly in "real" forms, but rather always use a molecule perhaps defined in `ui-forms`, so a realistic form could look like:

```js-jsx
  <InputField label="" help="" value="" />
  <InputField label="" help="" value="" valid />

<FieldSet stacked>
  <RadioField label="" value="" />
  <RadioField label="" value="" />
  <RadioField label="" value="" />
</FieldSet>

<InputField label="" help="" value="" error />
<CheckboxField label="" value="" help="" />
```

Reply:
I see, we were talking about the same, with the only difference that you want the `FormControl` to be used explicitly, while I thought that the `FormControl` is wrapped around the individual children of `FieldSet` implicitly.

But I agree with the explicit use of `FormControl` here:
* It allows us to have exceptions that won't have to be handled by the `FormControl`
* As you said, the combination of `FieldSet` and `FormControl` will most likely used by `ui-forms` rather than apps, leaving the option for external apps/devs to use their own way

Reply:
Also (interesting to @HendrikThePendric as well):

With the current implementation, we have no 'advanced' control over the layout inside the `FieldSet`.

For a `RadioGroup`/`CheckboxGroup` for example, we'd also need the `Help` text inside the `FieldSet`, which right now - in the inline version - is not possible.

Reply:
> It is awkward that the FieldSet adds spacing to form components.

I think this is quite a good point, however, this thread is making me think that we actually need to stop looking at the code and the components we have at this point in time. Instead we should have a broader discussion about the components we need to a) layout form-fields, and b) build inputs that are composed of several components.

To explain why I think this, I'll give a short summary on how we got here in the first place:
1. I was working on the `RadioGroup` component in `ui-forms`, and I realised it needed some styled elements:
    1. Some sort of text label to go above the radio inputs, with some margin
    1. Some spacing between the Radios, and support for stacked and inline
    1. I needed some sort of wrapper component, so the each form component in `ui-forms` would have a root component with some spacing.
1. We decided `ui-forms` should contain no styling, so we should introduce helper components in `ui-core`. Still fully on board with that.
1. Some of these helper components have gone through a full-on Pokemon evolution by now. We started with a `RadioGroup` (which was meh), which then evolved into a `ToggleGroup` (better, but still had quite a narrow scope and the `ToggleGroup.Label` component was awkward), and that then evolved into the `FieldSet` component.
1. This `FieldSet` component makes a lot of sense, and is nice and generic. And the fact that such a generic component currently adds spacing to form components is indeed fishy.... So I would agree we should probably remove this behaviour.....
1. However.... Now we've gotten to a point where we have a nice new `FieldSet` component, which makes a lot of sense conceptually, but I don't have a use-case for at the moment. And at the same time, I am still not able to compose the `ui-forms` `RadioGroup` from a bunch of `ui-core` components, which was how this all started. So we have sort of come full circle...

I think the summary above highlights that we are currently still working out what we need. So perhaps we'd better focus on that first, and continue building stuff later.

Do you agree @varl @Mohammer5 ?

Reply:
>  Now we've gotten to a point where we have a nice new FieldSet component, which makes a lot of sense conceptually, but I don't have a use-case for at the moment.

`fielset` itself is a very useful component: It enables accessibility in forms.
See: http://usability.com.au/2013/04/accessible-forms-1-labels-and-identification/

So I think we need

1. FieldSet (which removes the default fieldset and legend styles)
1. FormControl (which wraps 1 field row for vertical spacing; could be 1 or more fields)
1. StackControl (which wraps 2 or more components; can be `vertical` or `horizontal`)

Reply:
Agreed, taking a step back and thinking about how to lay out the components is a good idea. My notes above are intended to hammer home the API design with concrete examples so we don't dig the same hole over and over again.

Reply:
> 5\. However.... Now we've gotten to a point where we have a nice new `FieldSet` component, which makes a lot of sense conceptually, but I don't have a use-case for at the moment. And at the same time, I am still not able to compose the `ui-forms` `RadioGroup` from a bunch of `ui-core` components, which was how this all started. So we have sort of come full circle...

Does not `FieldSet` + `FormControl` give you ability to compose the `ui-forms::RadioGroup` component?

Reply:
> Does not FieldSet + FormControl give you ability to compose the ui-forms::RadioGroup component?

I'll interfere her:
It would probably, yes. But then we couldn't use the `FormControl` for other inputs in `ui-forms` to group inputs from `ui-core` with `Help` from `u-core`, which is why I proposed the 3 components above.

With those 3 you could also have multiple radio groups together in 1 FieldSet, each RadioGroup separated by a `FormControl`